agent/utils/dockerapi_compare_versions.go

```diff
@@ -0,0 +1,127 @@
+	minor             int
+}
+
+// DockerVersionMatches returns whether or not a version matches a given selector.
+// The selector can be any of the following:
+//
+// * x.y -- Matches a version exactly the same as the selector version
+// * >=x.y -- Matches a version greater than or equal to the selector version
+// * >x.y -- Matches a version greater than the selector version
+// * <=x.y -- Matches a version less than or equal to the selector version
+// * <x.y -- Matches a version less than the selector version
+// * x.y,a.b -- Matches if the version matches either of the two selector versions
+func (lhs DockerVersion) Matches(selector string) (bool, error) {
+	lhsVersion, err := parseDockerVersions(string(lhs))
+	if err != nil {
+		return false, err
+	}
+
+	if strings.Contains(selector, ",") {
+		orElements := strings.Split(selector, ",")
+		for _, el := range orElements {
+			if elMatches, err := lhs.Matches(el); err != nil {
+				return false, err
+			} else if elMatches {
+				return true, nil
+			}
+		}
+		// No elements matched
+		return false, nil
+	}
+
+	if strings.HasPrefix(selector, ">=") {
+		rhsVersion, err := parseDockerVersions(selector[2:])
+		if err != nil {
+			return false, err
+		}
+		return compareDockerVersions(lhsVersion, rhsVersion) >= 0, nil
+	} else if strings.HasPrefix(selector, ">") {
+		rhsVersion, err := parseDockerVersions(selector[1:])
+		if err != nil {
+			return false, err
+		}
+		return compareDockerVersions(lhsVersion, rhsVersion) > 0, nil
+	} else if strings.HasPrefix(selector, "<=") {
+		rhsVersion, err := parseDockerVersions(selector[2:])
+		if err != nil {
+			return false, err
+		}
+		return compareDockerVersions(lhsVersion, rhsVersion) <= 0, nil
+	} else if strings.HasPrefix(selector, "<") {
+		rhsVersion, err := parseDockerVersions(selector[1:])
+		if err != nil {
+			return false, err
+		}
+		return compareDockerVersions(lhsVersion, rhsVersion) < 0, nil
+	}
+
+	rhsVersion, err := parseDockerVersions(selector)
+	if err != nil {
+		return false, err
+	}
+	return compareDockerVersions(lhsVersion, rhsVersion) == 0, nil
+}
+
+func parseDockerVersions(version string) (dockerapi_version, error) {
+	var result dockerapi_version
+	// x.x
+	versionParts := strings.Split(version, ".")
+	// [x, x]
+	if len(versionParts) != 2 {
+		return result, fmt.Errorf("Not enough '.' characters in the version part")
+	}
+	major, err := strconv.Atoi(versionParts[0])
+	if err != nil {
+		return result, fmt.Errorf("Cannot parse major version as int: %v", err)
+	}
+	minor, err := strconv.Atoi(versionParts[1])
+	if err != nil {
+		return result, fmt.Errorf("Cannot parse minor version as int: %v", err)
+	}
+	result.major = major
+	result.minor = minor
+	return result, nil
+}
+
+// compareDockerVersions compares two docker api versions, 'lhs' and 'rhs', and returns -1 if lhs is less
+// than rhs, 0 if they are equal, and 1 if lhs is greater than rhs
+func compareDockerVersions(lhs, rhs dockerapi_version) int {
```

Comment:
For readability, we can use a switch as described here: https://github.com/golang/go/wiki/Switch#missing-expression

Reply:
Changed to Switch